File: algorithms/2_medium/M/03922_minimum_flips_to_make_binary_string_coherent/solution.c
static inline size_t min_size(const size_t a, const size_t b)
{
    return a < b ? a : b;
}

int minFlips(const char* s)
{
    assert(s != NULL && s[0] != '\0');
    size_t ones = 0;
    size_t zeros = 0;
    char last_char = ' ';

    for (size_t i = 0; s[i] != '\0'; i += 1) {
        switch (s[i]) {
            case '0':
                zeros += 1;
                break;
            case '1':
                ones += 1;
                break;
            default:
                fprintf(stderr, "unreachable!(), invalid input");
                exit(1);
        }
        last_char = s[i];
    }

    if (zeros == 0 || ones <= 1) {
        return 0;
    }

    const size_t extra_keep_one = (s[0] == '1' && last_char == '1') ? 2 : 1;
    const size_t res = min_size(zeros, ones - extra_keep_one);

    assert(res <= (size_t)INT_MAX);

    return (int)res;
}

File: algorithms/2_medium/M/03922_minimum_flips_to_make_binary_string_coherent/solution.rs
impl Solution {
    pub fn min_flips(s: String) -> i32 {
        let bytes = s.as_bytes();
        assert!(bytes.len() > 0);
        let (mut ones, mut zeros) = (0_usize, 0_usize);

        for &b in bytes {
            match b {
                b'0' => zeros += 1,
                b'1' => ones += 1,
                _ => unreachable!(),
            }
        }

        if zeros == 0 || ones <= 1 {
            return 0;
        }

        let extra_keep_one = if bytes[0] == b'1' && *bytes.last().unwrap() == b'1' { 2 } else { 1 };

        zeros.min(ones - extra_keep_one).try_into().unwrap()
    }
}
